apps/dcos_dns/src/dcos_dns_mesos.erl

```diff
@@ -0,0 +1,384 @@
+
+-spec(handle_task_updated(task_id(), task(), state()) -> state()).
+handle_task_updated(TaskId, Task, #state{tasks=Tasks}=State) ->
+    {Tasks0, Ops} = task_updated(TaskId, Task, Tasks),
+    handle_push_ops(Ops, State#state{tasks=Tasks0}).
+
+-spec(task_updated(task_id(), task(), Tasks) -> {Tasks, Ops}
+    when Tasks :: #{task_id() => [dns:dns_rr()]},
+         Ops :: [riak_dt_orswot:orswot_op()]).
+task_updated(TaskId, Task, Tasks) ->
+    TaskState = maps:get(state, Task),
+    case {?IS_RUNNING(TaskState), maps:is_key(TaskId, Tasks)} of
+        {Same, Same} ->
+            {Tasks, []};
+        {false, true} ->
+            {Records, Tasks0} = maps:take(TaskId, Tasks),
+            {Tasks0, [{remove_all, Records}]};
+        {true, false} ->
+            TaskRecords = task_records(TaskId, Task),
+            Tasks0 = maps:put(TaskId, TaskRecords, Tasks),
+            {Tasks0, [{add_all, TaskRecords}]}
+    end.
+
+-spec(task_records(#{task_id() => task()}) -> #{task_id() => [dns:dns_rr()]}).
+task_records(Tasks) ->
+    maps:fold(fun (TaskId, #{state := TaskState} = Task, Acc) ->
+        case ?IS_RUNNING(TaskState) of
+            true ->
+                TaskRecords = task_records(TaskId, Task),
+                maps:put(TaskId, TaskRecords, Acc);
+            false -> Acc
+        end
+    end, #{}, Tasks).
+
+-spec(task_records(task_id(), task()) -> dns:dns_rr()).
+task_records(_TaskId, Task) ->
+    lists:flatten([
+        task_agentip(Task),
+        task_containerip(Task),
+        task_autoip(Task)
+    ]).
+
+-spec(task_agentip(task()) -> [dns:dns_rr()]).
+task_agentip(#{name := Name, framework := Fwrk, agent_ip := AgentIP}) ->
+    DName = format_name([Name, Fwrk, <<"agentip">>], ?DCOS_DOMAIN),
+    dns_records(DName, [AgentIP]).
+
+-spec(task_containerip(task()) -> [dns:dns_rr()]).
+task_containerip(#{name := Name, framework := Fwrk, task_ip := TaskIPs}) ->
+    DName = format_name([Name, Fwrk, <<"containerip">>], ?DCOS_DOMAIN),
+    dns_records(DName, TaskIPs);
+task_containerip(_Task) ->
+    [].
+
+-spec(task_autoip(task()) -> [dns:dns_rr()]).
+task_autoip(#{name := Name, framework := Fwrk,
+              agent_ip := AgentIP, task_ip := TaskIPs} = Task) ->
+    %% if task.port_mappings then agent_ip else task_ip
+    DName = format_name([Name, Fwrk, <<"autoip">>], ?DCOS_DOMAIN),
+    Ports = maps:get(ports, Task, []),
+    dns_records(DName,
+        case lists:any(fun is_port_mapping/1, Ports) of
+            true -> [AgentIP];
+            false -> TaskIPs
+        end
+    ).
+
+-spec(is_port_mapping(dcos_net_mesos_state:task_port()) -> boolean()).
+is_port_mapping(#{host_port := _HPort}) ->
+    true;
+is_port_mapping(_Port) ->
+    false.
+
+%%%===================================================================
+%%% Masters functions
+%%%===================================================================
+
+-spec(handle_masters(state()) -> state()).
+handle_masters(#state{masters=MRRs}=State) ->
```

Comment:
are these new records that we are adding? Because currently, we do not have master record with "dcos" zoneName

Reply:
See https://github.com/dcos/dcos-net/pull/37#discussion_r181501582